Running CMake throws linking error - cannot find gcc

Following the instructions in the README, we create a [docker image](https://github.com/prashanthr05/sample-docker-conda-cpp-project/blob/main/dockerfiles/Dockerfile) with ubuntu 22.04 installing basic packages like `wget`, `git `, and `ca-certificates` from [apt](https://github.com/prashanthr05/sample-docker-conda-cpp-project/blob/main/deps/apt-packages.txt) and the rest of the required packages for C++ development using mambaforge given in [environment.yml](https://github.com/prashanthr05/sample-docker-conda-cpp-project/blob/main/deps/environment.yml). We copy the source code files within the project directory in a workspace created in home.

CMake configuration step  throws the following error,

``` bash
-- The CXX compiler identification is GNU 10.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /home/user/conda/envs/myenv/bin/x86_64-conda-linux-gnu-c++
-- Check for working CXX compiler: /home/user/conda/envs/myenv/bin/x86_64-conda-linux-gnu-c++ - broken
CMake Error at /home/user/conda/envs/myenv/share/cmake-3.24/Modules/CMakeTestCXXCompiler.cmake:62 (message):
  The C++ compiler

    "/home/user/conda/envs/myenv/bin/x86_64-conda-linux-gnu-c++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/user/my_ws/sample-cpp/build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/home/user/conda/envs/myenv/bin/make -f Makefile cmTC_9116f/fast && /home/user/conda/envs/myenv/bin/make  -f CMakeFiles/cmTC_9116f.dir/build.make CMakeFiles/cmTC_9116f.dir/build
    make[1]: Entering directory '/home/user/my_ws/sample-cpp/build/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_9116f.dir/testCXXCompiler.cxx.o
    /home/user/conda/envs/myenv/bin/x86_64-conda-linux-gnu-c++   -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/user/conda/envs/myenv/include  -o CMakeFiles/cmTC_9116f.dir/testCXXCompiler.cxx.o -c /home/user/my_ws/sample-cpp/build/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    Linking CXX executable cmTC_9116f
    /home/user/conda/envs/myenv/bin/cmake -E cmake_link_script CMakeFiles/cmTC_9116f.dir/link.txt --verbose=1
    /home/user/conda/envs/myenv/bin/x86_64-conda-linux-gnu-c++ -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/user/conda/envs/myenv/include  -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/user/conda/envs/myenv/lib -Wl,-rpath-link,/home/user/conda/envs/myenv/lib -L/home/user/conda/envs/myenv/lib  CMakeFiles/cmTC_9116f.dir/testCXXCompiler.cxx.o -o cmTC_9116f 
    /home/user/conda/envs/myenv/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: cannot find -lgcc: No such file or directory
    collect2: error: ld returned 1 exit status
    make[1]: *** [CMakeFiles/cmTC_9116f.dir/build.make:99: cmTC_9116f] Error 1
    make[1]: Leaving directory '/home/user/my_ws/sample-cpp/build/CMakeFiles/CMakeTmp'
    make: *** [Makefile:127: cmTC_9116f/fast] Error 2
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:6 (project)


-- Configuring incomplete, errors occurred!
See also "/home/user/my_ws/sample-cpp/build/CMakeFiles/CMakeOutput.log".
See also "/home/user/my_ws/sample-cpp/build/CMakeFiles/CMakeError.log".
```

>  For an example of the cleaning commands used by the "official" mambaforge image, check https://github.com/conda-forge/miniforge-images/blob/ac6652a64d584f2dd97365febe3994037db1226c/ubuntu/Dockerfile#L35 .

In fact, I was actually using the instructions from that file for installing `mambaforge` inside the container and cleaning up the installation to keep the image lightweight

But what was actually causing the problem was this line,

https://github.com/prashanthr05/sample-docker-conda-cpp-project/blob/0403ee04332c4ca7960b46dcbabfa68a5104770c/dockerfiles/Dockerfile#L68

where I was deleting the static libraries which clearly reflects the linking error  `cannot find -lgcc`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added sample C++ project in a Docker-containerized Conda environment.
 
 ### Changed
+- Fixed cleaning up static libraries in the mamba virtual environment causing linking errors during compilation (https://github.com/prashanthr05/sample-docker-conda-cpp-project/pull/2)
 
 
 ### Removed

--- a/dockerfiles/Dockerfile
+++ b/dockerfiles/Dockerfile
@@ -3,6 +3,9 @@ FROM ubuntu:22.04
 ARG username=user
 ARG userid=1000
 ARG groupid=100
+ARG virtualenvname=myenv
+ARG wsname=my_ws
+ARG cppprojname=sample-cpp
 
 SHELL ["/bin/bash", "-c"]
 
@@ -12,13 +15,16 @@ ENV DEBIAN_FRONTEND=noninteractive
 # Set the locale
 ENV LANG=C.UTF-8 LC_ALL=C.UTF-8
 
-# Create non-root user for safety
+# Set environment variables
 ENV USER $username
 ENV UID $userid
 ENV GID $groupid
 ENV HOME /home/$USER
-ENV VENVNAME myenv
+ENV VENVNAME $virtualenvname
+ENV WORKSPACE $wsname
+ENV CPPPROJECT $cppprojname
 
+# Create non-root user for safety
 RUN adduser --disabled-password \
     --gecos "Non-root User" \
     --uid ${UID} \
@@ -42,14 +48,15 @@ RUN chown $UID:$GID /tmp/environment.yml
 USER ${USER}
 
 # Create a workspace and a sample cpp project
-ENV WSDIR  $HOME/my_ws
-ENV SAMPLECPPDIR $WSDIR/sample-cpp
+ENV WSDIR  $HOME/$WORKSPACE
+ENV SAMPLECPPDIR $WSDIR/$CPPPROJECT
 RUN mkdir -p $SAMPLECPPDIR
 WORKDIR $SAMPLECPPDIR
 COPY CMakeLists.txt $SAMPLECPPDIR/CMakeLists.txt
 COPY src/main.cpp $SAMPLECPPDIR/src/main.cpp
 
 # Prepare conda environment
+# see https://github.com/conda-forge/miniforge-images/blob/ac6652a64d584f2dd97365febe3994037db1226c/ubuntu/Dockerfile
 ENV CONDA_DIR=$HOME/conda
 ENV PATH=$CONDA_DIR/bin:$CONDA_DIR/envs/$VENVNAME/bin:$PATH
 RUN  wget --no-hsts --quiet "https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-$(uname)-$(uname -m).sh" -O ~/mambaforge_installer.sh && \
@@ -58,16 +65,15 @@ RUN  wget --no-hsts --quiet "https://github.com/conda-forge/miniforge/releases/l
     mamba clean --tarballs --index-cache --packages --yes && \
     find ${CONDA_DIR} -follow -type f -name '*.a' -delete && \
     find ${CONDA_DIR} -follow -type f -name '*.pyc' -delete && \
-    mamba clean --force-pkgs-dirs --all --yes && \
+    mamba clean --all --yes && \
     mamba create -n ${VENVNAME} python=3.9 && \
     echo ". ${CONDA_DIR}/etc/profile.d/conda.sh && conda activate ${VENVNAME}" >> ~/.bashrc
 
 RUN mamba env update --name $VENVNAME \
                      --file /tmp/environment.yml && \
     mamba clean --tarballs --index-cache --packages --yes && \
-    find ${CONDA_DIR} -follow -type f -name '*.a' -delete && \
     find ${CONDA_DIR} -follow -type f -name '*.pyc' -delete && \
-    mamba clean --force-pkgs-dirs --all --yes
+    mamba clean --all --yes
 
 
 

